baybe/recommenders/pure/bayesian/base.py

```diff
@@ -47,75 +54,53 @@ def _validate_deprecated_argument(self, _, value) -> None:
             )
 
-    def _setup_botorch_acqf(
-        self,
-        searchspace: SearchSpace,
-        train_x: Optional[pd.DataFrame] = None,
-        train_y: Optional[pd.DataFrame] = None,
+    def _transform_data(
+        self, data: pd.DataFrame, searchspace: SearchSpace, objective: Objective
     ) -> None:
-        """Create the current acquisition function from provided training data.
+        """Transform data from experimental to computational representation."""
+        self._train_x = searchspace.transform(data)
+        self._train_y = objective.transform(data)
+
+    def _setup_botorch_acqf(self, searchspace: SearchSpace) -> None:
+        """Create the acquisition function from the current training data.
 
         The acquisition function is stored in the private attribute
         ``_acquisition_function``.
 
         Args:
             searchspace: The search space in which the experiments are to be conducted.
-            train_x: The features of the conducted experiments.
-            train_y: The corresponding response values.
-
-        Raises:
-            NotImplementedError: If the setup is attempted from empty training data
         """
-        if train_x is None or train_y is None:
-            raise NotImplementedError(
-                "Bayesian recommenders do not support empty training data yet."
-            )
-
-        best_f = train_y.max().item()
-        surrogate_model = self._fit(searchspace, train_x, train_y)
+        best_f = self._train_y.max().item()
+        self.surrogate_model.fit(searchspace, *to_tensor(self._train_x, self._train_y))
         self._botorch_acqf = self.acquisition_function.to_botorch(
-            surrogate_model, best_f
+            self.surrogate_model, best_f
         )
 
-    def _fit(
-        self,
-        searchspace: SearchSpace,
-        train_x: pd.DataFrame,
-        train_y: pd.DataFrame,
-    ) -> Surrogate:
-        """Train a fresh surrogate model instance.
-
-        Args:
-            searchspace: The search space.
-            train_x: The features of the conducted experiments.
-            train_y: The corresponding response values.
-
-        Returns:
-            A surrogate model fitted to the provided data.
-
-        Raises:
-            ValueError: If the training inputs and targets do not have the same index.
-        """
-        # validate input
-        if not train_x.index.equals(train_y.index):
-            raise ValueError("Training inputs and targets must have the same index.")
-
-        self.surrogate_model.fit(searchspace, *to_tensor(train_x, train_y))
-
-        return self.surrogate_model
-
     def recommend(  # noqa: D102
         self,
         searchspace: SearchSpace,
+        objective: Optional[Objective],
         batch_size: int = 1,
-        train_x: Optional[pd.DataFrame] = None,
-        train_y: Optional[pd.DataFrame] = None,
+        train_df: Optional[pd.DataFrame] = None,
```

Comment:
Yes, exactly. Needs to be `Optional` from an interface perspective, otherwise we have an LSP violation (and mypy will also complain accordingly). Since we expose it as `Optional` via the `RecommenderProtocol` and further set `None` as default in `PureRecommender`, the subclasses now have to mirror this. Hence, we handle that case internally in the method, throwing an error when we encounter it. Makes sense / better ideas?

Reply:
No objections, so I'll resolve :D